### PR TITLE
[SPARK-48444][SQL][TESTS] Use withSQLConf in tests: Refactor SQLQuerySuite

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Use withSQLConf in tests when it is appropriate.

### Why are the changes needed?

Enforce good practice for setting config in test cases.

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

Existing UT

### Was this patch authored or co-authored using generative AI tooling?

NO

### CODE REVIEW

**Comment by amaliujia** [comment]
@cloud-fan

**Review by cloud-fan — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 1]
```diff
@@ -1108,35 +1110,31 @@ abstract class SQLQuerySuiteBase extends QueryTest with SQLTestUtils with TestHi
...
-        """
-          |insert into table dynparttest1 partition(pdate)
-          | select count(*), cast('2015-05-21' as date) as pdate from src
-        """.stripMargin)
-      checkAnswer(
-        sql("select * from dynparttest1"),
-        Seq(Row(500, java.sql.Date.valueOf("2015-05-21"))))
+    withTable("dynparttest1", "dynparttest2") {
+      withSQLConf("hive.exec.dynamic.partition.mode" -> "strict") {
+        sql("set hive.exec.dynamic.partition.mode=nonstrict")
```
Comment (yaooqinn) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:1115]: This behave different now？

Comment (amaliujia) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: nice catch. updated

Comment (yaooqinn) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: It still seems wrong to me. You should test with an inner wrapper using `withSQLConf + nonstrict` and restore the `strict` mode with an outer wrapper `try-finally`.

Comment (amaliujia) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: That is the purpose of using `withSQLConfig`. So each test case should set the configs it needs and restores only to default values.

A test case should not pollute other test cases nor care about starting values of confs of other test cases.

Comment (yaooqinn) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: So replacing `nonstrict` with `strict` is an intentional change?

Comment (amaliujia) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: So previous code does something like
1. set the mode
2. the recover it by the finally so other test cases are not impacted.

The new code also set the mode, and recover it by `withSQLConf`

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala [hunk 2]
```diff
@@ -1108,35 +1110,30 @@ abstract class SQLQuerySuiteBase extends QueryTest with SQLTestUtils with TestHi
...
-      sql(
-        """
-          |insert into table dynparttest1 partition(pdate)
-          | select count(*), cast('2015-05-21' as date) as pdate from src
-        """.stripMargin)
-      checkAnswer(
-        sql("select * from dynparttest1"),
-        Seq(Row(500, java.sql.Date.valueOf("2015-05-21"))))
+    withTable("dynparttest1", "dynparttest2") {
+      withSQLConf("hive.exec.dynamic.partition.mode" -> "strict") {
```
Comment (yaooqinn) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala:1114]: shouldn't it be nonstrict?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: https://github.com/apache/spark/pull/46778#discussion_r1618216246 move the following questions there

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala (hunk 2 above)
Comment (amaliujia) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: oops this is right. yes

**Review by yaooqinn — APPROVED** [review]

**Pull request closed without merging**